File: 22FALL/74.search-a-2-d-matrix.py
#
# @lc app=leetcode id=74 lang=python3
#
# [74] Search a 2D Matrix
#

# @lc code=start
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        left, right = 0, m*n-1
        while left <= right:
            mid = (left+right)//2
            i, j = mid//n, mid%n
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] < target:
                left = mid + 1
            else:
                right = mid -1
        return False
# @lc code=end

File: VSCode_CPP/74.search-a-2-d-matrix.cpp
/*
 * @lc app=leetcode id=74 lang=cpp
 *
 * [74] Search a 2D Matrix
 */

// @lc code=start
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size(), n=matrix[0].size();
        int i=0, j=0;
        while(i<m && j<n) {
            if (matrix[i][j]>target) return false;
            else if (matrix[i][j] == target) return true;
            else if (i+1<m && matrix[i+1][j]<=target) i++;
            else j++;
        }
        return false;
    }
};
// @lc code=end

